parsing/procq.ml

```diff
@@ -8,43 +8,194 @@
+
+let gstate () = (!state).current_state
+
+let get_keyword_state () = (gstate()).kwstate
+
+let terminal s = CLexer.terminal (get_keyword_state()) s
+
+let reset_to_base state = {
+  base_state = state.base_state;
+  current_state = state.base_state;
+  current_sync_extensions = [];
+  user_state = GramState.empty;
+  custom_entries = EntryDataMap.empty;
+}
+
+let modify_state_unsync f state =
+  let is_base = state.base_state == state.current_state in
+  let base_state = f state.base_state in
+  let current_state = if is_base then base_state else f state.current_state in
+  { state with base_state; current_state }
+
+let modify_state_unsync f () = state := modify_state_unsync f !state
+
+let add_keyword_tok tok =
+  modify_state_unsync (fun {estate;kwstate} ->
+      let kwstate = CLexer.add_keyword_tok kwstate tok in
+      {estate; kwstate})
+    ()
+
+let make_entry_unsync make remake state =
+  let is_base = state.base_state == state.current_state in
+  let base_estate, e = make state.base_state.estate in
+  let base_state = { state.base_state with estate = base_estate } in
+  let current_state = if is_base then base_state else
+      let current_estate = remake state.current_state.estate e in
+      { state.current_state with estate = current_estate }
+  in
+  { state with base_state; current_state }, e
+
+let make_entry_unsync make remake () =
+  let statev, e = make_entry_unsync make remake !state in
+  state := statev;
+  e
+
+let add_kw = { add_kw = CLexer.add_keyword_tok }
+
+let epsilon_value (type s tr a) f (e : (s, tr, a) Symbol.t) =
+  let r = Production.make (Rule.next Rule.stop e) (fun x _ -> f x) in
+  let { GState.estate; kwstate } = gstate() in
+  let estate, entry = Entry.make "epsilon" estate in
+  let ext = Fresh (Gramlib.Gramext.First, [None, None, [r]]) in
+  let estate, kwstate = safe_extend add_kw estate kwstate entry ext in
+  let strm = Stream.empty () in
+  let strm = Parsable.make strm in
+  try Some (Entry.parse entry strm {estate;kwstate}) with e when CErrors.noncritical e -> None
+
+let extend_gstate {GState.kwstate; estate} e ext =
+  let estate, kwstate = safe_extend add_kw estate kwstate e ext in
+  {GState.kwstate; estate}
+
+(* XXX rename to grammar_extend_unsync? *)
+let grammar_extend e ext =
+  let extend_one g = extend_gstate g e ext in
+  modify_state_unsync extend_one ()
+
+(** NB: [extend_statement =
+         gram_position option * single_extend_statement list]
+    and [single_extend_statement =
+         string option * gram_assoc option * production_rule list]
+    and [production_rule = symbol list * action]
+
+    In [single_extend_statement], first two parameters are name and
+    assoc iff a level is created *)
+
+(** Type of reinitialization data *)
+type gram_reinit = Gramlib.Gramext.g_assoc * Gramlib.Gramext.position
+
+type extend_rule =
+| ExtendRule : 'a Entry.t * 'a extend_statement -> extend_rule
+| ExtendRuleReinit : 'a Entry.t * gram_reinit * 'a extend_statement -> extend_rule
```

Comment:
NB the reinit is now unused, but cleaning it up can wait until CI confirms this approach can work